Source/misdat.cpp

```diff
@@ -127,68 +129,162 @@ MissileDataStruct MissileData[] = {
+	{ "ex_ora1",  MFILE_EXORA1,     1,       MFLAG_HIDDEN,            0,            13,                                     96,                    -12               },
+	{ "ex_blu3",  MFILE_EXBL3,      1,       MFLAG_HIDDEN,            0,             7,                                    292,                    114               },
+	{ "",         MFILE_NONE,       0,       0                                                                                                                       },
 	// clang-format on
 };
 
+MisFileData::MisFileData(const char *name, uint8_t animName, uint8_t animFAmt, uint32_t flags,
+			 std::array<uint8_t, 16> animDelay, std::array<uint8_t, 16> animLen,
+			 std::array<int16_t, 16> animWidth, std::array<int16_t, 16> animWidth2)
+	: name(name)
+	, animName(animName)
+	, animFAmt(animFAmt)
+	, flags(flags)
+	, animDelay(animDelay)
+	, animLen(animLen)
+	, animWidth(animWidth)
+	, animWidth2(animWidth2)
+{
+	if (flags & MFLAG_ALLOW_SPECIAL)
+		pinnedMem.reserve(1);
+	else
+		pinnedMem.reserve(animFAmt);
+}
+
+namespace  {
+
+template <typename T>
+constexpr std::array<T, 16> duplicate(T value, unsigned count)
+{
+	std::array<T, 16> ret = {};
+	for (unsigned i = 0; i < count; i++)
+		ret[i] = value;
+	return ret;
+}
+
+} //namespace
+
+MisFileData::MisFileData(const char *name, uint8_t animName, uint8_t animFAmt, uint32_t flags,
+			 uint8_t animDelay, uint8_t animLen,
+			 int16_t animWidth, int16_t animWidth2)
+	: MisFileData(name, animName, animFAmt, flags,
+		      duplicate<uint8_t>(animDelay, animFAmt), duplicate<uint8_t>(animLen, animFAmt),
+		      duplicate<int16_t>(animWidth, animFAmt), duplicate<int16_t>(animWidth2, animFAmt))
+{
+}
+
+MisFileData::MisFileData(const char *name, uint8_t animName, uint8_t animFAmt, uint32_t flags,
+			 uint8_t animDelay, std::array<uint8_t, 16> animLen,
+			 int16_t animWidth, int16_t animWidth2)
+	: MisFileData(name, animName, animFAmt, flags,
+		      duplicate<uint8_t>(animDelay, animFAmt), animLen,
+		      duplicate<int16_t>(animWidth, animFAmt), duplicate<int16_t>(animWidth2, animFAmt))
+{
+}
+
+MisFileData::MisFileData(const char *name, uint8_t animName, uint8_t animFAmt, uint32_t flags,
+			 std::array<uint8_t, 16> animDelay, uint8_t animLen,
+			 int16_t animWidth, int16_t animWidth2)
+	: MisFileData(name, animName, animFAmt, flags,
+		      animDelay, duplicate<uint8_t>(animLen, animFAmt),
+		      duplicate<int16_t>(animWidth, animFAmt), duplicate<int16_t>(animWidth2, animFAmt))
+{
+}
+
+MisFileData::MisFileData(const char *name, uint8_t animName, uint8_t animFAmt, uint32_t flags,
+			 uint8_t animDelay, std::array<uint8_t, 16> animLen,
+			 std::array<int16_t, 16> animWidth, std::array<int16_t, 16> animWidth2)
+	: MisFileData(name, animName, animFAmt, flags,
+		      duplicate<uint8_t>(animDelay, animFAmt), animLen,
+		      animWidth, animWidth2)
+{
+}

```

Comment:
While this helps to simplify the data table it ads ~70 lines of logic and only covers 5/8 of the possible permutations, meaning that some one editing the data table could easily end up with build errors for what might not be obvious to them.

Reply:
I've come up with a much simpler solution: extending std::array and adding a constructor that fills the entire array with something.

Reply:
It takes care of all permutations without requiring extra MisFileData constructors.